src/main/java/com/squareup/javawriter/builders/CodeWriter.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2015 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.javawriter.builders;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSortedMap;
+import com.squareup.javawriter.ClassName;
+import com.squareup.javawriter.StringLiteral;
+import com.squareup.javawriter.TypeName;
+import com.squareup.javawriter.TypeNames;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Converts a {@link JavaFile} to a string suitable to both human- and javac-consumption. This
+ * honors imports, indentation, and deferred variable names.
+ */
+class CodeWriter {
+  private final String indent = "  ";
+  private final StringBuilder out;
```

Comment:
The original choice to code to `Appendable` and `CharSequence` was quite intentional.  Ridiculously enough, we have builds that will actually feel some pain from not being able to write directly to the file objects and having to turn `Name`s into `String`s.